samples/mqtt/basic_pub_sub/main.cpp

```diff
@@ -309,20 +301,29 @@ int main(int argc, char *argv[])
         /*
          * Subscribe for incoming publish messages on topic.
          */
-        auto onSubAck = [&](Mqtt::MqttConnection &, uint16_t packetId, const String &topic, Mqtt::QOS, int errorCode) {
-            if (packetId)
-            {
-                fprintf(stdout, "Subscribe on topic %s on packetId %d Succeeded\n", topic.c_str(), packetId);
-            }
-            else
-            {
-                fprintf(stdout, "Subscribe failed with error %s\n", aws_error_debug_str(errorCode));
-            }
-            conditionVariable.notify_one();
-        };
+        std::promise<bool> subscribeFinishedPromise;
+        auto onSubAck =
+            [&](Mqtt::MqttConnection &, uint16_t packetId, const String &topic, Mqtt::QOS QoS, int errorCode) {
+                if (errorCode)
+                {
+                    fprintf(stderr, "Subscribe failed with error %s\n", aws_error_debug_str(errorCode));
+                }
+                else
+                {
+                    if (!packetId || QoS == AWS_MQTT_QOS_FAILURE)
+                    {
+                        fprintf(stderr, "Subscribe rejected by the broker.");
+                    }
+                    else
+                    {
+                        fprintf(stdout, "Subscribe on topic %s on packetId %d Succeeded\n", topic.c_str(), packetId);
+                    }
+                }
+                subscribeFinishedPromise.set_value(true);
+            };
 
         connection->Subscribe(topic.c_str(), AWS_MQTT_QOS_AT_LEAST_ONCE, onPublish, onSubAck);
-        conditionVariable.wait(uniqueLock);
+        subscribeFinishedPromise.get_future().wait();
```

Comment:
We don't do anything differently if the subscribe failed ... which is odd
probably should print/exit from the callback when we get the error_code, or exit after the wait() if the value returned is false